python/cudf/cudf/utils/cudautils.py

```diff
@@ -113,20 +113,30 @@ def gpu_copy(inp, out):
 
 
 def astype(ary, dtype):
-    if ary.dtype == np.dtype(dtype):
+
+    src_dtype = ary.dtype
+    dst_dtype = np.dtype(dtype)
+
+    if src_dtype == dst_dtype:
         return ary
-    elif (
-        ary.dtype == np.int64 or ary.dtype == np.dtype("datetime64[ms]")
-    ) and (dtype == np.dtype("int64") or dtype == np.dtype("datetime64[ms]")):
-        return ary.view(dtype)
-    elif ary.size == 0:
-        return rmm.device_array(shape=ary.shape, dtype=dtype)
-    else:
-        out = rmm.device_array(shape=ary.shape, dtype=dtype)
-        if out.size > 0:
-            configured = gpu_copy.forall(out.size)
-            configured(ary, out)
-        return out
+
+    src_is_datetime = np.issubdtype(src_dtype, np.datetime64)
+    dst_is_datetime = np.issubdtype(dst_dtype, np.datetime64)
+
+    if src_is_datetime or dst_is_datetime:
+        # if converting to/from a datetime and to/from int64,
+        # it's safe to return a zero-copy view
+        if (src_dtype if dst_is_datetime else dst_dtype) == 'int64':
+            return ary.view(dst_dtype)
+
+    if ary.size == 0:
+        return rmm.device_array(shape=ary.shape, dtype=dst_dtype)
+
+    out = rmm.device_array(shape=ary.shape, dtype=dst_dtype)
+    if out.size > 0:
+        configured = gpu_copy.forall(out.size)
+        configured(ary, out)
+    return out
```

Comment:
This whole function is in process of being reworked to use `gdf_cast` anyway, so I'd pay attention to that function as well.